storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -55,6 +113,2304 @@ public class UnifiedLog {
+                                latestTimestampSegment = segment;
+                            }
+                        }
+                        // cache the timestamp and offset
+                        TimestampOffset maxTimestampSoFar = latestTimestampSegment.readMaxTimestampAndOffsetSoFar();
+                        // lookup the position of batch to avoid extra I/O
+                        OffsetPosition position = latestTimestampSegment.offsetIndex().lookup(maxTimestampSoFar.offset);
+                        Optional<FileRecords.TimestampAndOffset> timestampAndOffsetOpt = findFirst(
+                                latestTimestampSegment.log().batchesFrom(position.position),
+                                item -> item.maxTimestamp() == maxTimestampSoFar.timestamp)
+                                    .flatMap(batch -> batch.offsetOfMaxTimestamp()
+                                        .map(offset -> new FileRecords.TimestampAndOffset(
+                                            batch.maxTimestamp(),
+                                            offset,
+                                            Optional.of(batch.partitionLeaderEpoch()).filter(epoch -> epoch >= 0))));
+                        return new OffsetResultHolder(timestampAndOffsetOpt);
+                    } else {
+                        // We need to search the first segment whose largest timestamp is >= the target timestamp if there is one.
+                        if (remoteLogEnabled() && !isEmpty()) {
+                            if (remoteOffsetReader.isEmpty()) {
+                                throw new KafkaException("RemoteLogManager is empty even though the remote log storage is enabled.");
+                            }
+
+                            AsyncOffsetReadFutureHolder<OffsetResultHolder.FileRecordsOrError> asyncOffsetReadFutureHolder =
+                                    remoteOffsetReader.get().asyncOffsetRead(topicPartition(), targetTimestamp,
+                                    logStartOffset, leaderEpochCache, () -> searchOffsetInLocalLog(targetTimestamp, localLogStartOffset()));
+                            return new OffsetResultHolder(Optional.empty(), Optional.of(asyncOffsetReadFutureHolder));
+                        } else {
+                            return new OffsetResultHolder(searchOffsetInLocalLog(targetTimestamp, logStartOffset));
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Checks if the log is empty.
+     * @return Returns True when the log is empty. Otherwise, false.
+     */
+    public boolean isEmpty() {
+        return logStartOffset == logEndOffset();
+    }
+
+    private Optional<FileRecords.TimestampAndOffset> searchOffsetInLocalLog(long targetTimestamp, long startOffset) {
+        // Cache to avoid race conditions.
+        List<LogSegment> segmentsCopy;
+        synchronized (lock) {
+            segmentsCopy = List.copyOf(logSegments());
```

Comment:
We can get rid of the locking and copying here since `logSegments()` does that now.